base/applications/taskmgr/graphctl.c

```diff
@@ -12,666 +12,363 @@
-    /*  clear out the existing garbage, re-start with a clean plot */
-    GraphCtrl_InvalidateCtrl(this, FALSE);
-}
-
-#if 0
-void TGraphCtrl::SetXUnits(const char* string)
+void TmgcDispose(PTM_GRAPH_CONTROL inst)
 {
-    strncpy(m_strXUnitsString, string, sizeof(m_strXUnitsString) - 1);
-    /*  clear out the existing garbage, re-start with a clean plot */
-    InvalidateCtrl();
-}
-
-void TGraphCtrl::SetYUnits(const char* string)
-{
-    strncpy(m_strYUnitsString, string, sizeof(m_strYUnitsString) - 1);
-    /*  clear out the existing garbage, re-start with a clean plot */
-    InvalidateCtrl();
-}
-#endif
-
-void GraphCtrl_SetGridColor(TGraphCtrl* this, COLORREF color)
-{
-    this->m_crGridColor = color;
-    /*  clear out the existing garbage, re-start with a clean plot */
-    GraphCtrl_InvalidateCtrl(this, FALSE);
-}
+    if (inst->PointBuffer)
+        HeapFree(GetProcessHeap(), 0, inst->PointBuffer);
 
-void GraphCtrl_SetPlotColor(TGraphCtrl* this, int plot, COLORREF color)
-{
-    this->m_crPlotColor[plot] = color;
-    DeleteObject(this->m_penPlot[plot]);
-    this->m_penPlot[plot] = CreatePen(PS_SOLID, 0, this->m_crPlotColor[plot]);
-    /*  clear out the existing garbage, re-start with a clean plot */
-    GraphCtrl_InvalidateCtrl(this, FALSE);
-}
+    DeleteObject(inst->hPenGrid);
+    DeleteObject(inst->hPen0);
+    DeleteObject(inst->hPen1);
+    DeleteObject(inst->hBrushBack);
+    DeleteObject(inst->hbmGraph);
+    DeleteObject(inst->hdcGraph);
 
-void GraphCtrl_SetBackgroundColor(TGraphCtrl* this, COLORREF color)
-{
-    this->m_crBackColor = color;
-    DeleteObject(this->m_brushBack);
-    this->m_brushBack = CreateSolidBrush(this->m_crBackColor);
-    /*  clear out the existing garbage, re-start with a clean plot */
-    GraphCtrl_InvalidateCtrl(this, FALSE);
+    return;
 }
 
-void GraphCtrl_InvalidateCtrl(TGraphCtrl* this, BOOL bResize)
+void TmgcAddPoint(PTM_GRAPH_CONTROL inst, BYTE val0, BYTE val1)
 {
-    /*  There is a lot of drawing going on here - particularly in terms of  */
-    /*  drawing the grid.  Don't panic, this is all being drawn (only once) */
-    /*  to a bitmap.  The result is then BitBlt'd to the control whenever needed. */
-    int i;
-    int nCharacters;
-    //int nTopGridPix, nMidGridPix, nBottomGridPix;
-
-    HPEN oldPen;
-    HPEN solidPen = CreatePen(PS_SOLID, 0, this->m_crGridColor);
-    /* HFONT axisFont, yUnitFont, oldFont; */
-    /* char strTemp[50]; */
-
-    /*  in case we haven't established the memory dc's */
-    /* CClientDC dc(this); */
-    HDC dc = GetDC(this->m_hParentWnd);
-
-    /*  if we don't have one yet, set up a memory dc for the grid */
-    if (this->m_dcGrid == NULL)
+    HDC    hdcg;
+    PBYTE  t;
+    RECT   rcDirt;
+
+    UINT   Prev0, Prev1, RetainingWidth;
+    INT    PrevY, CurrY, p, v;
+
+    hdcg = inst->hdcGraph;
+    RetainingWidth = inst->BitmapWidth - PLOT_SHIFT;
+    t = inst->PointBuffer;
```

Comment:
It is known PointBuffer can be NULL (see checks above), but here you use and dereference it without checks? (see below)

Reply:
if PointBuffer is NULL (that only be the case if HeapAlloc failed to allocate a pretty small amount of memory (not bigger, than 1 page), this function won't be called, because we do PostQuitMessage on this fail before.